NoSuchFileException if fcrepo 4 export didn't create directories alongside ACL files

I ran a successful export of a fcrepo 4.7.5 using the import/export tool version 0.3.0, but when trying to migrate the export to fcrepo 5 I run into a NoSuchFileException:

>java.lang.RuntimeException: java.nio.file.NoSuchFileException: archivo_fcrepo4.7.5_export/fedora/rest/af/21/ab/71/af21ab71-c690-4c3b-a608-a76dd1529044
	at org.fcrepo.upgrade.utils.F47ToF5UpgradeManager.convertAcl(F47ToF5UpgradeManager.java:372)
	at org.fcrepo.upgrade.utils.F47ToF5UpgradeManager.lambda$upgradeRdfAndCreateHeaders$5(F47ToF5UpgradeManager.java:262)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.fcrepo.upgrade.utils.F47ToF5UpgradeManager.upgradeRdfAndCreateHeaders(F47ToF5UpgradeManager.java:200)
	at org.fcrepo.upgrade.utils.F47ToF5UpgradeManager.processFile(F47ToF5UpgradeManager.java:149)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at org.fcrepo.upgrade.utils.F47ToF5UpgradeManager.processDirectory(F47ToF5UpgradeManager.java:114)
	at org.fcrepo.upgrade.utils.F47ToF5UpgradeManager.start(F47ToF5UpgradeManager.java:108)
	at org.fcrepo.upgrade.utils.UpgradeUtilDriver.run(UpgradeUtilDriver.java:75)
	at org.fcrepo.upgrade.utils.UpgradeUtilDriver.main(UpgradeUtilDriver.java:60)
Caused by: java.nio.file.NoSuchFileException: archivo_fcrepo4.7.5_export/fedora/rest/af/21/ab/71/af21ab71-c690-4c3b-a608-a76dd1529044
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:171)
	at java.base/sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
	at java.base/java.nio.file.Files.readAttributes(Files.java:1853)
	at java.base/java.nio.file.FileTreeWalker.getAttributes(FileTreeWalker.java:220)
	at java.base/java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:277)
	at java.base/java.nio.file.FileTreeWalker.walk(FileTreeWalker.java:323)
	at java.base/java.nio.file.FileTreeIterator.<init>(FileTreeIterator.java:71)
	at java.base/java.nio.file.Files.walk(Files.java:3898)
	at java.base/java.nio.file.Files.walk(Files.java:3953)
	at org.fcrepo.upgrade.utils.F47ToF5UpgradeManager.convertAcl(F47ToF5UpgradeManager.java:347)
	... 19 more

From what I can tell the ACL that is trying to be converted is empty.  It does not contain any child nodes.  Other ACLs that were processed before this error did contain children.  Because they had children they had directories created for them alongside the turtle file.  The ones that didn't have children did not have directories created which I believe led to the above error.  I wrote a quick script to create these empty directories for all ACLs that were missing them and the migration ran successfully.  My guess is that there needs to be a check added before attempting the `File.walk` to skip if the directory doesn't exist.
https://github.com/fcrepo-exts/fcrepo-upgrade-utils/blob/b2c43e1f6b822a5d594501b958917725c77baaf9/src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java#L347

--- a/README.md
+++ b/README.md
@@ -38,6 +38,9 @@ usage: java -jar fcrepo-upgrade-util-<version>.jar
                                      export. Default: Turtle
  -R,--resource-info-file <arg>       The path of the file that contains a
                                      list of resources to be processed
+    --skip-acls                      Skip creating fcr:acl resources
+                                     and migrate webac:Acl and
+                                     acl:Authorization resources normally
  -s,--source-version <arg>           The version of Fedora that was the
                                      source of the export. Valid values:
                                      5+,4.7.5

--- a/src/main/java/org/fcrepo/upgrade/utils/Config.java
+++ b/src/main/java/org/fcrepo/upgrade/utils/Config.java
@@ -41,6 +41,7 @@ public class Config {
     private String fedoraUserAddress = DEFAULT_USER_ADDRESS;
     private boolean forceWindowsMode = false;
     private Path resourceInfoFile;
+    private boolean skipAcls = false;
 
     // S3 Options
     private boolean writeToS3;
@@ -256,6 +257,20 @@ public class Config {
     }
 
     /**
+     * @return true if should skip special handling of acl resources
+     */
+    public boolean isSkipAcls() {
+        return skipAcls;
+    }
+
+    /**
+     * @param skipAcls true if should skip special handling of acl resources
+     */
+    public void setSkipAcls(final boolean skipAcls) {
+        this.skipAcls = skipAcls;
+    }
+
+    /**
      * @return true if should write migrated resources to S3
      */
     public boolean isWriteToS3() {
@@ -395,6 +410,7 @@ public class Config {
                 ", fedoraUser='" + fedoraUser + '\'' +
                 ", fedoraUserAddress='" + fedoraUserAddress + '\'' +
                 ", forceWindowsMode=" + forceWindowsMode +
+                ", skipAcls=" + skipAcls +
                 ", writeToS3=" + writeToS3 +
                 ", s3Region='" + s3Region + '\'' +
                 ", s3Endpoint='" + s3Endpoint + '\'' +

--- a/src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java
+++ b/src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java
@@ -181,7 +181,7 @@ class F47ToF5UpgradeManager extends UpgradeManagerBase implements UpgradeManager
 
         //skip if ACL or Authorization: these files are upgraded through a separate code path
         // see convertAcl() below.
-        if (rdfTypes.contains(ACL) || rdfTypes.contains(AUTHORIZATION)) {
+        if (!this.config.isSkipAcls() && (rdfTypes.contains(ACL) || rdfTypes.contains(AUTHORIZATION))) {
             newLocation.toFile().delete();
             return;
         }
@@ -250,7 +250,7 @@ class F47ToF5UpgradeManager extends UpgradeManagerBase implements UpgradeManager
                     }
                 }
                 binaryHeaders.put(CONTENT_TYPE_HEADER, Collections.singletonList(mimetype));
-            } else if (statement.getPredicate().equals(ACCESS_CONTROL)) {
+            } else if (!this.config.isSkipAcls() && statement.getPredicate().equals(ACCESS_CONTROL)) {
                 //remove the current statement across both past versions and latest version
                 model.remove(currentStatement);
                 rewriteModel.set(true);

--- a/src/main/java/org/fcrepo/upgrade/utils/UpgradeUtilDriver.java
+++ b/src/main/java/org/fcrepo/upgrade/utils/UpgradeUtilDriver.java
@@ -168,6 +168,8 @@ public class UpgradeUtilDriver {
         config.setFedoraUser(cmd.getOptionValue("migration-user"));
         config.setFedoraUserAddress(cmd.getOptionValue("migration-user-address"));
 
+        config.setSkipAcls(cmd.hasOption("skip-acls"));
+
         config.setWriteToS3(cmd.hasOption("write-to-s3"));
         config.setS3Region(cmd.getOptionValue("s3-region"));
         config.setS3Endpoint(cmd.getOptionValue("s3-endpoint"));
@@ -280,6 +282,12 @@ public class UpgradeUtilDriver {
                 .hasArg(true)
                 .desc("The address of the user OCFL versions are attributed to. Default: "
                         + Config.DEFAULT_USER_ADDRESS)
+                .required(false)
+                .build());
+        configOptions.addOption(Option.builder()
+                .longOpt("skip-acls")
+                .hasArg(false)
+                .desc("Skip creating fcr:acl resources and migrate webac:Acl and acl:Authorization resources normally")
                 .required(false)
                 .build());
 

--- a/src/test/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManagerTest.java
+++ b/src/test/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManagerTest.java
@@ -5,6 +5,7 @@
  */
 package org.fcrepo.upgrade.utils;
 
+import static org.fcrepo.upgrade.utils.RdfConstants.ACCESS_CONTROL;
 import static org.fcrepo.upgrade.utils.RdfConstants.AUTHORIZATION;
 import static org.fcrepo.upgrade.utils.RdfConstants.FEDORA_LAST_MODIFIED_DATE;
 import static org.junit.Assert.assertEquals;
@@ -155,7 +156,7 @@ public class F47ToF5UpgradeManagerTest {
                                                .filter(x -> x.getPredicate().equals(FEDORA_LAST_MODIFIED_DATE))
                                                .findFirst().get();
         assertTrue("There should be a last modified date", lastModifiedStatement != null);
-        assertEquals("The subject should be be the acl: ",
+        assertEquals("The subject should be the acl: ",
                      "http://localhost:8080/rest/container1/fcr:acl",
                      lastModifiedStatement.getSubject().getURI());
     }
@@ -248,7 +249,6 @@ public class F47ToF5UpgradeManagerTest {
             }
         }
 
-
         //validate acl
         //ensure there are two authorizations under the hash uri #auth0 and #auth1
         final var model = RdfUtil.parseRdf(Path.of(output.toString(), "rest/container1/fcr%3Aacl.nt"), Lang.TTL);
@@ -272,6 +272,72 @@ public class F47ToF5UpgradeManagerTest {
         assertEquals("The subject should be be the acl: ",
                      "http://localhost:8080/rest/container1/fcr:acl",
                      lastModifiedStatement.getSubject().getURI());
+    }
+
+    @Test
+    public void testUpgradeSkipAcls() throws Exception {
+        //prepare
+        final File tmpDir = tempFolder.newFolder();
+        final File input = new File(TARGET_DIR + "/test-classes/4.7.5-export");
+        final File output = new File(tmpDir, "output");
+        output.mkdir();
+
+        final var config = new Config();
+        config.setSourceVersion(FedoraVersion.V_4_7_5);
+        config.setTargetVersion(FedoraVersion.V_5);
+        config.setInputDir(input);
+        config.setOutputDir(output);
+	config.setSkipAcls(true);
+        //run
+        UpgradeManager upgradeManager = UpgradeManagerFactory.create(config);
+        upgradeManager.start();
+        //ensure all expected files exist
+        final String[] expectedFiles =
+            new String[]{"rest.ttl",
+                         "rest.ttl.headers",
+                         "rest/acl.ttl",
+                         "rest/acl/authZ1.ttl",
+                         "rest/acl/authZ2.ttl",
+                         "rest/external1",
+                         "rest/external1/fcr%3Ametadata.ttl",
+                         "rest/container1.ttl",
+                         "rest/container1.ttl.headers",
+                         "rest/container1/fcr%3Aversions/20201015053947.ttl",
+                         "rest/container1/fcr%3Aversions/20201015053947.ttl.headers",
+                         "rest/container1/fcr%3Aversions/20201015053526.ttl",
+                         "rest/container1/fcr%3Aversions/20201015053526.ttl.headers",
+                         "rest/container1/testbinary.binary",
+                         "rest/container1/testbinary/fcr%3Ametadata.ttl",
+                         "rest/container1/testbinary.binary.headers",
+                         "rest/container1/testbinary/fcr%3Ametadata/fcr%3Aversions/20201015053717.ttl",
+                         "rest/container1/testbinary/fcr%3Ametadata/fcr%3Aversions/20201015053717.ttl.headers",
+                         "rest/container1/testbinary/fcr%3Ametadata/fcr%3Aversions/20201015053848.ttl",
+                         "rest/container1/testbinary/fcr%3Ametadata/fcr%3Aversions/20201015053848.ttl.headers",
+                         "rest/container1/testbinary/fcr%3Aversions/20201015053848.binary",
+                         "rest/container1/testbinary/fcr%3Aversions/20201015053848.binary.headers",
+                         "rest/container1/testbinary/fcr%3Aversions/20201015053717.binary",
+                         "rest/external1.external.headers",
+                         "rest/external1.external"};
+
+        for (String f : expectedFiles) {
+            assertTrue(f + " does not exist as expected", new File(output, f).exists());
+        }
+
+        final String[] unexpectedFiles =
+            new String[]{"rest/container1/fcr%3Aacl.ttl"};
+
+        for (String f : unexpectedFiles) {
+            assertFalse(f + " should not exist.", new File(output, f).exists());
+        }
+
+        //validate acl special handling was skipped
+        //ensure access control statement is still present in the container
+	//acl and authorization resources were migrated and checked in expectedFiles
+        final var model = RdfUtil.parseRdf(Path.of(output.toString(), "rest/container1.ttl"), Lang.TTL);
+        final var accessControlStatement = model.listStatements().toList().stream()
+                                                .filter(x -> x.getPredicate().equals(ACCESS_CONTROL))
+                                                .findFirst().get();
+        assertTrue("There should still be an access control statement", accessControlStatement != null);
     }
 
     private Map<String, List<String>> deserializeHeaders(final File headerFile) throws IOException {

